include/ametsuchi/buffer.inc

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+template <typename T>
+CircularStack<T>::CircularStack(size_t s) : cap(s), sz(0), i_end(0) {
+  if (s == 0) throw Exception("Buffer size cannot be zero");
+  v = (T *)malloc(sizeof(T) * capacity());
+}
+
+template <typename T>
+CircularStack<T>::CircularStack(CircularStack &&cs) {
+  v = cs.v;
+  cap = cs.cap;
+  sz = cs.sz;
+  i_end = cs.i_end;
+  cs.v = nullptr;
+}
+
+template <typename T>
+CircularStack<T>::~CircularStack() {
+  if (v) free(v);
+  v = nullptr;
+}
+
+template <typename T>
+void CircularStack<T>::resize(size_t s) {
+  cap = s;
+  sz = std::min(capacity(), size());
+  v = (T *)realloc(v, sizeof(T) * capacity());
+}
+
+template <typename T>
+void CircularStack<T>::push(T &&t) {
+  v[i_end] = std::move(t);
+  i_end = (i_end + 1) % capacity();
+  if (size() < capacity()) sz++;
+}
+
+template <typename T>
+void CircularStack<T>::push(const T &t) {
+  v[i_end] = t;
+  i_end = (i_end + 1) % capacity();
+  if (size() < capacity()) sz++;
+}
+
+template <typename T>
+void CircularStack<T>::pop(size_t n) {
+  sz = size() < n ? 0 : size() - n;
```

Comment:
It Should add `i_end = ( i_end + capacity() -1 ) % capacity()`.
And `diff()` should change to 
```
constexpr size_t diff(size_t p, size_t n) const {
  return p < n ? p + capacity() - n : p - n;
}
```
This is failed following that
```
  TEST(CirularStack, PushAndPop) {
    auto size = 10;
    std::stack<int> st;
    CircularStack<uint64_t> cs(size);
    for(auto i = 0; i < size-1; i++ ) {
      st.push(i);
      cs.push(i);
    }
    for(auto i=0; i < size-1; i++ ) {
      ASSERT_EQ( st.top(), cs[cs.size()-1] );
      st.pop(); cs.pop();
    }
  }
```